src/CoreBundle/private/library/classes/TCMSTableEditor/TCMSTableFieldWriter.class.php

```diff
@@ -120,66 +136,118 @@ public function Save(&$postData, $bDataIsInSQLForm = false)
-            $currentFieldTypeRow = MySqlLegacySupport::getInstance()->fetch_assoc($currentFieldTypeResult);
-        }
-        $aOriginalData = $this->oTable->sqlData;
+        $this->oldData = $this->oTable->sqlData;
+        $this->oldTableName = $this->_oParentRecord->sqlData['name'];
+        $this->isChangeFromTable = $postData['bTargetTableChangeForMLTField'] ?? false; // true (from TCMSTableWriter) prohibits target changes
 
-        $newFieldTypeQuery = "SELECT * FROM `cms_field_type` WHERE `id` = '".MySqlLegacySupport::getInstance()->real_escape_string($postData['cms_field_type_id'])."'";
-        $newFieldTypeResult = MySqlLegacySupport::getInstance()->query($newFieldTypeQuery);
-        $newFieldTypeRow = MySqlLegacySupport::getInstance()->fetch_assoc($newFieldTypeResult);
+        return parent::Save($postData);
+    }
 
-        $oldName = $postData['name'];
-        if (array_key_exists('name', $this->oTable->sqlData)) {
-            $oldName = $this->oTable->sqlData['name'];
-        }
+    /**
+     * {@inheritdoc}
+     */
+    protected function PostSaveHook(&$oFields, &$oPostTable)
+    {
+        parent::PostSaveHook($oFields, $oPostTable);
+
+        $this->_oField = $this->adaptFieldAndRelatedTables($oPostTable->sqlData);
 
-        if ($bTypeChanged) {
-            $this->_oField->ChangeFieldTypePreHook();
+        $this->getAutoclassesCacheWarmer()->updateTableById($oPostTable->sqlData['cms_tbl_conf_id']);
+    }
+
+    /**
+     * @param array $postData
+     *
+     * @return TCMSField - the new definition after changes
+     *
+     * @throws DBALException
+     */
+    protected function adaptFieldAndRelatedTables(array $postData): TCMSField
+    {
+        $oldTypeId = $this->oldData['cms_field_type_id'];
+        $newTypeId = $postData['cms_field_type_id'];
+        $newName = $this->oTable->sqlData['name'];
+
+        $bTypeChanged = $oldTypeId !== $newTypeId;
+        $currentFieldTypeRow = $this->getFieldTypeDefinition($oldTypeId);
+        $newFieldTypeRow = $this->getFieldTypeDefinition($newTypeId);
+
+        /**
+         * @var $oldField TCMSField
+         */
+        $oldField = $this->oTablePreChangeData->GetFieldObject();
+        $oldField->oDefinition = $this->oTablePreChangeData;
+        $oldField->sTableName = $this->oldTableName;
+        $oldField->name = $this->oldData['name'];
+
+        if (true === $bTypeChanged) {
+            $oldField->ChangeFieldTypePreHook();
         }
-        if (!$bTargetTableChangeForMLTField) {
-            if ($this->_oField->AllowDeleteRelatedTablesBeforeFieldSave($postData, $currentFieldTypeRow, $newFieldTypeRow)) {
-                $this->_oField->DeleteRelatedTables();
-            } elseif ($this->_oField->AllowRenameRelatedTablesBeforeFieldSave($postData, $currentFieldTypeRow, $newFieldTypeRow)) {
-                $this->_oField->RenameRelatedTables($postData);
+
+        if (false === $this->isChangeFromTable) {
```

Comment:
Can we invert the `if` clause? Negated conditions together with `else` are hard to read.

Reply:
OK

Reply:
ok